pkg/ebpf/c/tracee.bpf.c

```diff
@@ -963,14 +1000,76 @@ BPF_HASH(module_init_map, u32, kmod_data_t, 256);                  // holds modu
 BPF_LRU_HASH(fd_arg_path_map, fd_arg_task_t, fd_arg_path_t, 1024); // store fds paths by task
 BPF_LRU_HASH(bpf_attach_map, u32, bpf_attach_t, 1024);             // holds bpf prog info
 BPF_LRU_HASH(bpf_attach_tmp_map, u32, bpf_attach_t, 1024);         // temporarily hold bpf_attach_t
+BPF_PERCPU_ARRAY(error_scratch, bpf_error_t, 1);                   // error scratch
+BPF_HASH(errors_count, bpf_error_t, bpf_error_count_t, 4096);      // errors count
+BPF_PERCPU_ARRAY(error_count_scratch, bpf_error_count_t, 1);       // error count scratch
+BPF_PERCPU_ARRAY(error_output_scratch, bpf_error_output_t, 1);     // error output scratch
 // clang-format on
 
 // EBPF PERF BUFFERS -------------------------------------------------------------------------------
 
+BPF_PERF_OUTPUT(errors, 1024);      // errors submission
 BPF_PERF_OUTPUT(events, 1024);      // events submission
 BPF_PERF_OUTPUT(file_writes, 1024); // file writes events submission
 BPF_PERF_OUTPUT(net_events, 1024);  // network events submission
 
+// HELPERS: ERRORS ---------------------------------------------------------------------------------
+
+static __always_inline void
+do_tracee_error(void *ctx, u32 opts, enum bpf_error_id id, s64 ret, u32 line, void *file)
+{
+    if (!ctx || !file)
+        return;
+
+    u32 zero = 0;
+    bpf_error_t *err = bpf_map_lookup_elem(&error_scratch, &zero);
+    if (unlikely(err == NULL))
+        return;
+
+    err->id = id;
+    err->ret = ret;
+    err->cpu = bpf_get_smp_processor_id();
+    err->line = line;
+    READ_KERN_STR_INTO(err->file, file);
+
+    bpf_error_count_t *counter_buf = bpf_map_lookup_elem(&error_count_scratch, &zero);
+    if (unlikely(counter_buf == NULL))
+        return;
```

Comment:
I don't see a use for this counter_buf.
You can simply save current timestamp as a local variable.

Reply:
Ok. I agree, considering it isn't a macro, so the memory footprint is low. I'll change it.

Reply:
Ah, the scratch is unneeded indeed. But the current timestamp can reside in a stack `counter_buf` as it needs to be used in the first error occurrence (map update).